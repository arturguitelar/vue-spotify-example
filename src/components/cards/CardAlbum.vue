<template>
  <div v-if="album" data-card="album" class="card card-album">
    <img :src="album.imageUrl || 'https://via.placeholder.com/500'" />
    <div class="info">
      <h3 class="name">{{ album.name }}</h3>
      <p class="artist">{{ album.artist }}</p>
    </div>
  </div>
  <div v-else data-card-message="empty">
    <p>Dados inv√°lidos</p>
  </div>
</template>

<script>
export default {
  name: "CardAlbum",
  props: {
    album: Object
  }
};
</script>

<style>
.card-album {
  display: flex;
  flex-direction: column;
}

.card-album img {
  width: 100%;
  height: auto;
}

.card-album .info {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  flex: auto;
  padding: 10px;
}

.card-album .name {
  font-size: 16px;
}

.card-album .artist {
  color: var(--gray-text);
  font-style: italic;
  font-size: 12px;
  text-align: right;
  border-top: var(--border-default);
  margin-top: 10px;
  padding-top: 10px;
}
</style>
