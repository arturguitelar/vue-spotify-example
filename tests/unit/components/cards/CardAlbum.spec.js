import { shallowMount } from "@vue/test-utils";
import CardAlbum from "@/components/cards/CardAlbum";

const DATASET_CARD = "[data-card='album']";
const DATASET_MESSAGE = "[data-card-message='empty']";
const IMAGE_URL = "https://via.placeholder.com/500";
const INVALID_DATA_MESSAGE = "Dados invÃ¡lidos";

describe("#CardAlbum", () => {
  describe("#FAIL", () => {
    test("render invalid data if doesn't receive correct object", () => {
      const wrapper = shallowMount(CardAlbum);

      expect(wrapper.find(DATASET_CARD).exists()).toBe(false);
      expect(wrapper.find(DATASET_MESSAGE).exists()).toBe(true);
      expect(wrapper.text()).toBe(INVALID_DATA_MESSAGE);
    });
  });

  describe("#SUCCESS", () => {
    const defaultData = {
      name: "Album name",
      imageUrl: IMAGE_URL,
      artist: "Artist name"
    };

    let wrapper;

    beforeEach(() => {
      wrapper = shallowMount(CardAlbum, {
        props: { album: defaultData }
      });
    });

    afterEach(() => {
      wrapper.unmount();
    });

    test("create a card with correct data", () => {
      expect(wrapper.vm.album).toStrictEqual(defaultData);
      expect(wrapper.props()).toEqual({ album: defaultData });
      expect(wrapper.find(DATASET_CARD).exists()).toBe(true);
      expect(wrapper.find(DATASET_MESSAGE).exists()).toBe(false);
    });

    test("image should be placed by placeholder if no image url", () => {
      defaultData.imageUrl = "";

      wrapper = shallowMount(CardAlbum, {
        props: { album: defaultData }
      });

      expect(wrapper.find("img").attributes()).toEqual({
        src: IMAGE_URL
      });
    });

    describe("Snapshots", () => {
      test("should render a card correctly", () => {
        expect(wrapper.wrapperElement).toMatchSnapshot();
      });
    });
  });
});
