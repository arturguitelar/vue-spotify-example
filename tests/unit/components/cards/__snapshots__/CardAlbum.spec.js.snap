// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#CardAlbum #SUCCESS Snapshots should render a card correctly 1`] = `
<div
  class="card card-album"
  data-card="album"
>
  <img
    src="https://via.placeholder.com/500"
  />
  <div
    class="info"
  >
    <h3
      class="name"
    >
      Album name
    </h3>
    <p
      class="artist"
    >
      Artist name
    </p>
  </div>
</div>
`;
